core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -818,6 +858,36 @@ private static AnnotatedValueResolver ofInjectableTypes0(String name, AnnotatedE
         };
     }
 
+    /**
+     * Returns an attribute resolver which retrieves a value specified by {@code attrKeys}
+     * from the {@link RequestContext}.
+     */
+    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
+    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
+        return (resolver, ctx) -> {
+            for (AttributeKey<?> attrKey : attrKeys) {
+                final Object value = ctx.context.attr(attrKey);
+                if (value != null) {
+
+                    final Class<?> rawContainerType = resolver.rawContainerType();
+                    final Class<?> elementType = resolver.elementType().isPrimitive() ?
+                                                 Primitives.wrap(resolver.elementType())
+                                                 : resolver.elementType();
+
+                    if ((rawContainerType != null && rawContainerType.isInstance(value)) ||
+                        (rawContainerType == null && elementType.isInstance(value))) {
+                        return value;
+                    }
+
+                    throw new ClassCastException("The value " + value +
+                                                 " cannot be cast to elemntType " +
+                                                 elementType.getSimpleName());
```

Comment:
- The method will be invoked whenever a request is received. Should we lazily create `elementType` when necessary?
```java
final boolean isValidType;
if (rawContainerType != null) {

} else {
   // call Primitives.wrap(resolver.elementType()) here
}
if (isValidType) {
     return value;
}
```
- It would be helpful for debugging if we provide `resolver.httpElementName` and `attrKey` in the error message.
- If the type is not matched with the value, should we scan until the last element?
```java
if (isValueType) {
   return value;
} else {
   lastValue = value;
   // continue to scan until finding a matched type.
   continue;
}

if (lastValue != null) {
    throw new ClassCastException(...);
}
```

Reply:
@ikhoon nim, thanks for your comment. 
Good point, i missed it 😢 

> The method will be invoked whenever a request is received. Should we lazily create elementType when necessary?

I modify codes to load `elementType()` laziliy. 
However, it still has problem to load `elementType()` duplicately when `AttrKeys.size() > 1` and iterate `AttrKey` more than 1. 

Any option is okay to me. 

IMHO, there is another option. 
```java
    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
        return (resolver, ctx) -> {
           final Class<?> rawContainerType = resolver.rawContainerType();
           final Class<?> elementType = .... ;
           for (AttributeKey<?> attrKey : attrKeys) { ... }
```


> It would be helpful for debugging if we provide resolver.httpElementName and attrKey in the error message.

Yes, i did apply it. 


> If the type is not matched with the value, should we scan until the last element?

Good point. i missed it. 
I'm done as you said. 
in this sequence i added `StringBuilder` to make error msg. 
Because `AttrKeys.size()` is bigger than 1, in that case, `value` is possible to be failed to cast more than 2.